Update the documentation about importing external certificates

davidfestal opened the pull request:
### What does this PR do?

This PR Updates the documentation about importing additional certificates in a Ce installation, to reflect the current status of Che on this Topic.

### What issues does this PR fix or reference?

This PR refers to issue https://issues.redhat.com/browse/RHDEVDOCS-2089

### Specify the version of the product this PR applies to.

This PR applies to the last Upstream release (7.18.2) and to the 2.4 downstream release. 

### PR Checklist

As the author of this Pull Request I made sure that:

- [x] `vale` has been run successfully against the PR branch
- [x] Link checker has been run successfully against the PR branch
- [ ] Documentation describes a scenario that is already covered by QE tests, otherwise an issue has been created and acknowledged by Che QE team
- [ ] Changed article references are updated where they are used (or a redirect has been set up on the docs side):
    - [ ] Dashboard [branding.constant.ts](https://github.com/eclipse/che-dashboard/blob/master/src/components/branding/branding.constant.ts) + [product.json](https://github.com/eclipse/che-dashboard/blob/master/src/assets/branding/product.json)
    - [ ] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/master/src/constants.ts)

davidfestal committed:
first changes

davidfestal committed:
Deeper update

davidfestal committed:
Cleanup

themr0c commented on modules/installation-guide/partials/proc_importing-tls-certificates-to-che-server-java-truststore.adoc:
```diff
@@ -1,92 +1,174 @@
 
 
 [id="importing-tls-certificates-to-{prod-id-short}-server-java-trustore_{context}"]
```
Can you also update the `id` and all `xref` statements pointing to this `id`?

davidfestal replied:
Done

tolusha commented on modules/installation-guide/partials/proc_importing-tls-certificates-to-che-server-java-truststore.adoc:
```diff
@@ -1,92 +1,174 @@
 ----
 +
-- Certificates are mounted in folder `/public-certs/` of the {prod-short} server container. This command returns the list of files in that folder:
+- {prod-short} mounts certificates in folder `/public-certs/` of the {prod-short} server container. This command returns the list of files in that folder:
```
I think that checking /public-certs folder is not enough.
when `ca-certs` configmap is updated then k8s automatically refreshes `/public-certs`.
but those certificates won't be added to java trust store of the che-server and keycloak.

l0rd commented on modules/installation-guide/partials/proc_importing-tls-certificates-to-che-server-java-truststore.adoc:
```diff
@@ -1,92 +1,174 @@
+Network communications between the components of a {prod-short} installation and the started workspaces, are all secured through the TLS protocol, and thus require the use of trusted certificate authorities. 
 
-To fix this problem, configure {prod-short} to authorize HTTPS communication with external services, such as identity and Git servers, by adding information about the TLS certificates to the {prod-short} configuration.
+In standard installations performed with the {prod-short} operator, the operator itself takes care of setting up and propagating the required TLS certificates, and no manual step should be necessary.
```
I would remove this sentence.

davidfestal replied:
Fixed

l0rd commented on modules/installation-guide/partials/proc_importing-tls-certificates-to-che-server-java-truststore.adoc:
```diff
@@ -1,92 +1,174 @@
-To fix this problem, configure {prod-short} to authorize HTTPS communication with external services, such as identity and Git servers, by adding information about the TLS certificates to the {prod-short} configuration.
+In standard installations performed with the {prod-short} operator, the operator itself takes care of setting up and propagating the required TLS certificates, and no manual step should be necessary.
+
+But in some cases it can be necessary to add TLS certificates to the {prod-short} installation, and have them used as trusted certificate authorities by {prod-short}.
```
```suggestion
In some cases it can be necessary to add TLS certificates to the {prod-short} installation, so that every {prod-short} component will consider them as signed by a trusted CA.
```

davidfestal replied:
Fixed

l0rd commented on modules/installation-guide/partials/proc_importing-tls-certificates-to-che-server-java-truststore.adoc:
```diff
@@ -1,92 +1,174 @@
+
+Typical cases that may require this addition are:
+
+* when using some specific configuration of the underlying Kubernetes or OpenShift cluster,
```
```suggestion
* when the underlying Kubernetes cluster uses TLS certificates signed by a CA that is not trusted,
```

davidfestal replied:
Fixed

l0rd commented on modules/installation-guide/partials/proc_importing-tls-certificates-to-che-server-java-truststore.adoc:
```diff
@@ -1,92 +1,174 @@
+Typical cases that may require this addition are:
+
+* when using some specific configuration of the underlying Kubernetes or OpenShift cluster,
+* to provide access to external secured services such as {identity-provider} or a Git server, from either the {prod-short} installation or the developer workspaces.
```
```suggestion
* when {prod-short} server or workspace components connect to external services such as {identity-provider} or a Git server that use TLS certificates signed by an untrusted CA.
```

davidfestal replied:
Fixed

davidfestal committed:
Add changes proposed by @l0rd

davidfestal committed:
Split the procedure into 2 sections:

according to whether the certificates are added;
- at installation time
- on an already-running installation.

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
+
+
+[id="importing-external-or-additional-tls-certificates_{context}"]
+= Importing external or additional TLS certificates to {prod-short}
```
Why "external or additional certificates"? Isn't "additional certificates" enough? Or even better "Importing untrusted TLS certificates to ...".

davidfestal replied:
Fixed

davidfestal replied:
in commit https://github.com/eclipse/che-docs/pull/1598/commits/ea6c439a0885f8b83daae5d6a262c694632af853

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
+[id="importing-external-or-additional-tls-certificates_{context}"]
+= Importing external or additional TLS certificates to {prod-short}
+
+Network communications between the components of a {prod-short} installation and the started workspaces, are all secured through the TLS protocol, and thus require the use of trusted certificate authorities. 
```
```suggestion
Internal communications between {prod-short} components are, by default, encrypted with TLS. Communications of {prod-short} components with external services like proxies, source code repositories, identity providers may require TLS tool. Those communications require the use of SSL certificates signed by trusted Certificate Authorities. 
```

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
+
+Network communications between the components of a {prod-short} installation and the started workspaces, are all secured through the TLS protocol, and thus require the use of trusted certificate authorities. 
+
+In some cases it can be necessary to add TLS certificates to the {prod-short} installation, so that every {prod-short} component will consider them as signed by a trusted CA.
```
```suggestion
When the certificates used by {prod-short} components or by an external service are signed by an untrusted CA it can be necessary to import the CA certificate in the {prod-short} installation, so that every {prod-short} component will consider them as signed by a trusted CA.
```

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
+* when {prod-short} server or workspace components connect to external services such as {identity-provider} or a Git server that use TLS certificates signed by an untrusted CA.
+
+To store those certificates, {prod-short} uses a dedicated ConfigMap. Its default name is `ca-certs` but {prod-short} allows configuring its name.
+On OpenShift, the {prod-short} operator even injects cluster trusted certificates into this ConfigMap automatically.
```
IMHO the OpenShift case should be a `[NOTE]` and should provide more details:
- link to latest documentation that explains how the certs are propagated
- the labels/annotation that the operator applies so that the propagation happen
Also consider that when this documentation is downstreamed OpenShift is the only platform. This means that a sentence like "On OpenShift..." doesn't make sense because it's always on OpenShift.

davidfestal replied:
What do you think of this:
![image](https://user-images.githubusercontent.com/686586/97467202-3e61cf00-1944-11eb-8983-76d7f9d01c98.png)

The link you see in this screenshot points to  https://docs.openshift.com/container-platform/4.4/networking/configuring-a-custom-pki.html#nw-proxy-configure-object_configuring-a-custom-pki

davidfestal replied:
Fixed in commit https://github.com/eclipse/che-docs/pull/1598/commits/3f481ae07d585dc12a78fe6300ba5daea7659c93

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
+* {prod-short} already uses some reserved file names to automatically inject certificates into the ConfigMap, so you should avoid using the following reserved file names to save your certificates:
+  ** `ca-bundle.crt`
+  ** `ca.crt`
+====
```
The paragraph above is repeated multiple times in this article. Can we move it to a different doc component and reference it multiple times?

davidfestal replied:
Fixed in commit https://github.com/eclipse/che-docs/pull/1598/commits/3f481ae07d585dc12a78fe6300ba5daea7659c93

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
+
+* The `{orch-cli}` tool is available.
+
+* You should first define the name of the ConfigMap you will use to import certificates:
```
```suggestion
* You should first gather the name of the ConfigMap used to import certificates:
```

davidfestal replied:
Fixed in commit https://github.com/eclipse/che-docs/pull/1598/commits/6d75512ae495960b96ee534fb64f7de41c0c91d6

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
+====
+endif::[]
+
+. Restart the {prod-short} operator, the {prod-short} server and {identity-provider} to load the new certificates:
```
I think that's not needed anymore after https://github.com/eclipse/che-operator/pull/487. You can specify that restart is not required on Che v7.20.0 or higher and CRW 2.5.0 or higher.

davidfestal replied:
Fixed in commit https://github.com/eclipse/che-docs/pull/1598/commits/3a2f0520b241716f87b2764864c4cfa7b8ac6fa2

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
+----
+$ openssl x509 -in __<certificate-file-path>__ -fingerprint -noout
+SHA1 Fingerprint=3F:DA:BF:E7:A7:A7:90:62:CA:CF:C7:55:0E:1D:7D:05:16:7D:45:60
+----
```
A verification at the workspace level: what config maps are created? what theia expects (an env variable? files mounted in a given folder? etc...)

davidfestal replied:
The problem is that as far as what I tested on CRW 2.4.0 release, the additional certificates added into the Che operator `ca-certs` config map **are not propagated to the workspace**: According to tests, and also according to [the code itself on the master branch](https://github.com/eclipse/che/blob/master/infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/Openshift4TrustedCAProvisioner.java#L81), the presence of a labelled `ca-certs` config map at the Che operator level, only triggers the creation of an **empty** `ca-certs` config map with the same annotation (used to inject OpenShift trusted CA bundles), but **without** the additional content that might have been manually added to the operator `ca-certs` config map.

So seems to completely invalidate the use of this config map to manage untrusted certificates that should be used also at the workspace level.

So maybe for now we should skip the workspace-related verification procedure and keep it for a future docs PR, when the propagation of the whole ca-certs CM content will be fixed.

@l0rd wdyt ?

l0rd replied:
But at least trusted CA bundle certs are injected in the cm right? That should be mentioned and we could add instructions to verify it. Please open an issue and notify deployment/platform teams.

davidfestal replied:
Verification steps at the workspace have been added in commit https://github.com/eclipse/che-docs/pull/1598/commits/96614f0174a3a113190f330001c439683a17d8db

l0rd reviewed (changes requested):
The document is much clearer now. Good job @davidfestal. But there are still some points to review.

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
+
+[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
+----
+$ get checluster {prod-checluster} -n __<{project-context}-namespace-name>__ -o json-path --jsonpath={.spec.server.serverTrustStoreConfigMapName}
```
```suggestion
$ {orch-cli} get checluster {prod-checluster} -n __<{project-context}-namespace-name>__ -o jsonpath={.spec.server.serverTrustStoreConfigMapName}
```

davidfestal replied:
Fixed in commit https://github.com/eclipse/che-docs/pull/1598/commits/89c462e804f3d5d1b9df27933ffac38dca7b8250

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
++
+[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
+----
+$ {orch-cli} rollout restart -n __<{project-context}-namespace-name>__ deployment/{prod-id-short}-operator
```
`oc rollout restart` is not a valid command. 
`oc rollout latest` works only with deployment config (not deployments) 

I am afraid that we should always use `kubectl` instead of `{orch-cli}` here.

davidfestal replied:
Fixed in commit https://github.com/eclipse/che-docs/pull/1598/commits/3a2f0520b241716f87b2764864c4cfa7b8ac6fa2

l0rd commented on modules/installation-guide/partials/proc_importing-external-or-additional-tls-certificates.adoc:
```diff
@@ -0,0 +1,218 @@
++
+[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
+----
+$ {orch-cli} exec -t __<{prod-id-short}-pod-name>__ -n {prod-namespace} -- keytool -list -keystore /home/che/cacerts
```
```suggestion
$ {orch-cli} exec -t __<{prod-id-short}-pod-name>__ -n {prod-namespace} -- keytool -list -keystore {prod-home}/cacerts
```

davidfestal replied:
Fixed in commit https://github.com/eclipse/che-docs/pull/1598/commits/3a2f0520b241716f87b2764864c4cfa7b8ac6fa2

davidfestal committed:
Suggested change by @l0rd

Co-authored-by: Mario Loriedo <mario.loriedo@gmail.com>

davidfestal committed:
@l0rd Suggested change by @l0rd

Co-authored-by: Mario Loriedo <mario.loriedo@gmail.com>

davidfestal committed:
Change the title according to @l0rd review comment

davidfestal committed:
Additional changes requested by @l0rd

davidfestal committed:
Small change asked by @l0rd

davidfestal committed:
2 additional changes requested by @l0rd

davidfestal committed:
Fix comment https://github.com/eclipse/che-docs/pull/1598#pullrequestreview-509615510

davidfestal committed:
Add verification steps at the workspace level

davidfestal committed:
Fix some of the vale errors

l0rd reviewed (approved):
LGTM Great job @davidfestal

Commit:
Merge branch 'master' into rhdevdocs-2089-reintegrate-external-ca-doc

yhontyk reviewed (approved):
LGTM
